### DISCUSSION TITLE
Two questions on pthread_start/end_main_hook:

### DISCUSSION THREAD

**Discussion by kunjsong01** [esbmc/esbmc#901] [Q&A] [answered]
Q1:
Where are they getting pulled in? Could anyone point me to the place in the source code please? 

![Screenshot 2023-02-10 at 11 12 28](https://user-images.githubusercontent.com/32592800/218082339-c089a3d5-e179-46f7-86a8-7d8ad7dfa161.png)

Q2: How are their symbol values getting populated? Could anyone point me to the place in the source code please? 

![Screenshot 2023-02-10 at 11 20 28](https://user-images.githubusercontent.com/32592800/218082385-66e4d309-d8cb-4681-bcfd-7047d2c921eb.png)

Many thanks!

**Comment by kunjsong01** [comment]
as for Q2, I traced it down to `add_cprover_library` in `clang_*_languaget::final` function: 
https://github.com/esbmc/esbmc/blob/3667d86d4fc343ee50f91fbd33955930006da554/src/clang-cpp-frontend/clang_cpp_language.cpp#L62-L65

But this is too top-level.

would be great if anyone could point me to the exact location where these symbols' values are getting populated... 

Thanks!

**Comment by mikhailramalho** [comment, marked as answer]
their models are located in pthread_lib.c.

They are pulled just like any other model in add_cprover_library. IIRC, we
collect all function symbols (only the empty ones?) in the program (and
their dependencies) and search if we got any model for them, from the
binary blob we generate with c2goto.

Em sex., 10 de fev. de 2023 às 07:32, KSong ***@***.***>
escreveu:

> Q1:
> Where are they getting pulled in? Could anyone point me to the place in
> the source code please?
>
> [image: Screenshot 2023-02-10 at 11 12 28]
> <https://user-images.githubusercontent.com/32592800/218082339-c089a3d5-e179-46f7-86a8-7d8ad7dfa161.png>
>
> Q2: How are their symbol values getting populated? Could anyone point me
> to the place in the source code please?
>
> [image: Screenshot 2023-02-10 at 11 20 28]
> <https://user-images.githubusercontent.com/32592800/218082385-66e4d309-d8cb-4681-bcfd-7047d2c921eb.png>
>
> Many thanks!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/discussions/901>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH4BXEG6X7OJJRKIFADWWYRNFANCNFSM6AAAAAAUXWCANM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 

Mikhail Ramalho.
